### ISSUE TITLE
OIDC RP does notHTTP Auth header as containing a valid OIDC id_token

### ISSUE THREAD

**Issue by barbj** [OpenLiberty/open-liberty#8808]
TS002246300
Our team is developing an application that should be able to accept OAuth 2.0 or OpenConnect ID tokens as Bearer tokens on the HTTP Authorization header to authenticate to the application hosted on WebSphere Liberty.

In addition, if the end user has not previously authenticated to our application, they are redirected to a Liberty-based Open ID Connect Provider for authentication. Then after they have authenticated, they are redirected back to our application as an OIDC RP.

In this latter scenario, if the openidConnectClient configuration sets the inboundPropagation attribute to "supported', this web-based authentication scenario does not work. Even though they have authenticated with the OIDC OP, they cannot access our application. The following error message appears in our application's log:

[5/13/19 11:42:59:273 PDT] 00044cdb .ibm.ws.security.openidconnect.client.jose4j.util.Jose4jUtil E CWWKS1712E: The OpenID Connect client [contentServicesUms] did not receive an ID token from the OpenID Connect provider [https://company.com:9443/oidc/endpoint/ums/token].

Below is the openidConnectClient configuration for our application:

```
<openidConnectClient id="ContentServicesUms"
scope="openid profile email"
responseType="id_token token"
clientId="myId"
clientSecret="mySecret"
issuerIdentifier="https://company.com:9443/oidc/endpoint/ums"
audiences="contentServicesUms"
validationMethod="introspect"
inboundPropagation="supported"
validationEndpointUrl="https://company.com:9443/oidc/endpoint/ums/introspect"
authorizationEndpointUrl="https://company.com:9443/oidc/endpoint/ums/authorize"
tokenEndpointUrl="https://company.com:9443/oidc/endpoint/ums/token">
</openidConnectClient>
```

If I change the inboundPropagation attribute to "none", then the web based authentication works, but the pre-existing OAuth 2.0 token and OIDC id_token scenarios no longer works when passed in as Authorization Bearer tokens.

I also noticed in the web based authentication flow, a cookie is set by the Liberty OIDC OP called "WASOidcCode" that contains the OIDC id_token.

When the responseType is changed to **code**, the scenario works.

Shouldn't a Liberty OIDC RP be able to use recognize this cookie as containing a valid OIDC id_token when using implicit flow (**responseType="id_token token"**)?

**Comment by brutif** [comment]
@barbj I tested this on 19007, looked at the code, and observed the following:

1) the responseType attribute documentation is confusing.  What it REALLY does is:   If set to code, the browser based flow uses an authorization code flow.  If set to anything else, the browser based flow will use an implicit flow.    Perhaps the way to think of it is it gets sent by the browser, informing the OP what kind of response it wants. 

2) the responseType attribute is unrelated to inbound propagation, at least on 19007. 

I can set it to "id_token token" and the web browser will get back both an access token and id token using the implicit flow.  And inbound propagation will work with both kinds of tokens.  Or I can set it to "code" and the web browser will get back an authorization code, and inbound propagation will continue to work with both kinds of tokens. 

3) Code flow is preferred for better security.  In the authorization code flow, the browser never sees the tokens, reducing any chance of them being stolen.  

I will update the metatype documentation on responseType, this will probably appear in 19010.

**Comment by brutif** [comment]
resolved by #8830
